src/liballoc/collections/vec_deque.rs

```diff
@@ -2709,6 +2709,33 @@ impl<T: fmt::Debug> fmt::Debug for VecDeque<T> {
 
 #[stable(feature = "vecdeque_vec_conversions", since = "1.10.0")]
 impl<T> From<Vec<T>> for VecDeque<T> {
+    /// Turn a `Vec<T>` into a `VecDeque<T>`.
+    ///
+    /// This avoids reallocating where possible, but the conditions for that are
+    /// strict, and subject to change, and so shouldn't be relied upon unless the
+    /// `Vec<T>` came from `From<VecDeque<T>>` and hasn't been reallocated.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::VecDeque;
+    ///
+    /// // Start with a `VecDeque<i32>`.
+    /// let deque: VecDeque<_> = (1..5).collect();
+    ///
+    /// // Turn it into a `Vec<i32>` with no allocation needed.
```

Comment:
What's the motivation for spending so much time worrying about reallocation here?

Reply:
I think there is a few open issues where people want to know about our allocation strategy for various `From` impls.  In other words, people want to know so they are being explicit about it.

Reply:
Basically, the only thing interesting about these two `From` impls is that they can (sometimes) avoid allocations compared to `.into_iter().collect()`.  So I want to write an example that shows the lack of reallocation, but I also want to write it in such a way that it's a situation where we'd be ok _committing_ to the lack of reallocation.  So I want to make an example that doesn't do things like "well, I happen to know that right now it doesn't reallocate if I `with_capacity(8)` and don't use all the capacity", because I don't want to write text elaborating why those conditions work.

I agree this example is far more complicated than most, though, so would be happy to change it.

Reply:
I don't know if an example is really helpful at all here, if we don't actually want to guarantee anything.

Reply:
@sfackler Ok, I removed the example from here.